### PR TITLE
GH-44098: [C++] Add home made _mm256_set_m128i for compilers who are missing it

### PR DESCRIPTION
**Pull request by zanmato1984**
### Rationale for this change

AVX2 intrinsic _mm256_set_m128i is missing in GCC versions <8.0 - this is a GCC bug discussed in [1], causing certain CI build failed.

[1] https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80582

### What changes are included in this PR?

Check the GCC version and use a home made replacement if necessary.

### Are these changes tested?

Manually tested.

### Are there any user-facing changes?

None.

* GitHub Issue: #44098

### CODE REVIEW

**Comment by zanmato1984** [comment]
cc @kou @assignUser @jonkeane

**Comment by zanmato1984** [comment]
@github-actions crossbow submit test-r-rstudio-r-base-4.1-opensuse155

File: cpp/src/arrow/util/simd.h [hunk 1]
```diff
@@ -27,18 +27,25 @@
 #else
 // gcc/clang (possibly others)
 
-#  if defined(ARROW_HAVE_BMI2)
+#  if defined(ARROW_HAVE_BMI2) || defined(ARROW_HAVE_RUNTIME_BMI2)
```
Comment (zanmato1984) [inline cpp/src/arrow/util/simd.h:30]: I try to find some common place to put simd specific code and found this one. However it seems this file is only adapting the static simd optimizations, thus the additions of the `ARROW_HAVE_RUNTIME*` which are for dynamic simd optimizations.

**Comment by zanmato1984** [comment]
Also cc @pitrou .

**Comment by pitrou** [comment]
@ursabot please benchmark lang=C++

**Comment by pitrou** [comment]
The so-called performance regressions are just flukes.

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp -g -r -g wheel

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp -g r -g wheel

**Comment by mapleFU** [comment]
Nice catch? Can we also try to try this to a gcc 7.5 and godbolt?

**Review by pitrou — APPROVED** [review]
+1, thank you @zanmato1984

**Pull request merged**